### ISSUE TITLE
Terraform detects spurious bucket policy changes before each apply after upgrading to 0.15.4

### ISSUE THREAD

**Issue by Smeb** [hashicorp/terraform-provider-aws#19627]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Relates #21968

### Terraform CLI and Terraform AWS Provider Version
Terraform v0.15.4
+ provider registry.terraform.io/hashicorp/aws v3.43.0

### Affected Resource(s)

* aws_s3_bucket_policy

### Description
I think this issue is likely related to #18878 . Essentially when using v0.15.4 of terraform, the new feature noting changes made outside of terraform displays a large number of buckets changed. However, the plan itself has no changes to apply.

After running `terraform apply` nothing happens. Using `terraform apply -refresh-only` also doesn't help. Every time we `plan` or `apply` we get the same (huge) list of changes. To fix this we just rolled back to using `0.15.3`.

### Expected Behavior

- The plan does not list any external changes to our infrastructure

### Actual Behavior

- The plan lists a large amount of external changes to our infrastructure, one for each bucket policy, despite the fact the whole account is only managed by terraform.

### Steps to Reproduce

I think this is somewhat state dependent - apologies for not providing an explicit reproduction case.

1. `terraform plan`

### References
* #18878

**Comment by sj26** [comment]
We're having this trouble too. The list of principals doesn't seem stably ordered, and is being compared with order sensitivity after upgrading for some reason.

**Comment by DavidGamba** [comment]
Affecting provider v3.45.0 and terraform v1.0.0 as well.

**Comment by sblask** [comment]
Same issue with the policy on `aws_kms_key`.

I can comment on #18878 anymore, but sorting the principals would probably not sort all the problems. There will probably be issues when using account IDs instead of ARNs as well. I now switched to:
```
    principals {
      identifiers = [for account in local.shared_accounts : "arn:aws:iam::${account}:root"]
      type        = "AWS"
    }
```
But:
```
    principals {
      identifiers = local.shared_accounts
      type        = "AWS"
    }
```
Is valid as well, but results in a diff like this:
```
                  ~ {
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::111111111111:root",
                              - "arn:aws:iam::222222222222:root",
                              - "arn:aws:iam::333333333333:root",
                              - "arn:aws:iam::444444444444:root",
                              - "arn:aws:iam::555555555555:root",
                              - "arn:aws:iam::666666666666:root",
                              - "arn:aws:iam::777777777777:root",
                              - "arn:aws:iam::888888888888:root",
                              + "777777777777",
                              + "222222222222",
                              + "888888888888",
                              + "333333333333",
                              + "111111111111",
                              + "666666666666",
                              + "555555555555",
                              + "111197162236",
                              + "444444444444",
                            ]
                        }
                        # (4 unchanged elements hidden)
                    },

```

**Comment by ewbankkit** [comment]
Potentially related: https://github.com/jen20/awspolicyequivalence/issues/10.

**Comment by YakDriver** [comment]
This was fixed with #21997
